rust/stackablectl/src/cmds/debug.rs

```diff
@@ -0,0 +1,380 @@
+                .as_ref()
+                .and_then(debug_container_status_of_pod(&debug_container_name))
+                .with_context(|| FindDebugContainerStatusSnafu {
+                    pod: pod_ref(),
+                    container: &self.container,
+                })?;
+            if let Some(termination) = debug_container_status
+                .last_state
+                .as_ref()
+                .and_then(|state| state.terminated.as_ref())
+            {
+                error!(
+                    error = termination.message,
+                    exit_code = termination.exit_code,
+                    "Debug container failed to start!"
+                );
+            }
+            info!("Attaching to container");
+            let mut attachment = pods
+                .attach(
+                    &self.pod,
+                    &AttachParams::interactive_tty().container(debug_container_name),
+                )
+                .await
+                .with_context(|_| AttachContainerSnafu {
+                    pod: pod_ref(),
+                    container: &self.container,
+                })?;
+            info!("Attached to container, if the shell line looks empty, press ENTER!");
+            let _raw = std::io::stdout()
+                .into_raw_mode()
+                .context(SetRawTtyModeSnafu)?;
+            futures::future::select_all([
+                update_terminal_size(
+                    attachment
+                        .terminal_size()
+                        .context(NoTerminalSizeChannelSnafu)?,
+                )
+                .boxed(),
+                tokio::io::copy(
+                    &mut attachment.stdout().context(NoStdoutChannelSnafu)?,
+                    &mut tokio::io::stdout(),
+                )
+                .map_ok(drop)
+                .context(ForwardStdoutSnafu)
+                .boxed(),
+                tokio::io::copy(
+                    &mut AsyncStdin::new()?,
+                    &mut attachment.stdin().context(NoStdinChannelSnafu)?,
+                )
+                .map_ok(drop)
+                .context(ForwardStdinSnafu)
+                .boxed(),
+            ])
+            .await
+            .0?;
+            Ok(String::new())
+        }
+        .instrument(span)
+        .await
+    }
+}
+
+fn generate_debug_container_name() -> String {
+    let mut rng = rand::thread_rng();
+    let mut name = "sble-debug-".to_string();
+    for _ in 0..5 {
+        name.push(rng.gen_range('a'..='z'));
+    }
+    name
+}
+
+/// Does true non-blocking reads of stdin, so that we can cancel properly on shutdown.
+/// The compromise is that it does not handle having things piped into it very well, since their write sides
+/// will also be turned non-blocking.
+///
+/// Only a single instance of AsyncStdin should exist at any one time.
+///
+/// Also, `AsyncStdin` will restore the mode of stdin when dropped.
+struct AsyncStdin {
```

Comment:
note: Aren't there any async Stdin implementations out there? If so, we could avoid maintaining our own implementation.

Reply:
tokio's and async_std's both use a worker thread that calls the blocking APIs. That's better for compatibility with pipes, but leads to a dangling prompt when shutting down.

Reply:
Termion also has `AsyncReader` (https://docs.rs/termion/latest/termion/struct.AsyncReader.html) but that doesn't seem to have a wakeness trigger, so we don't know when we need to poll.

Reply:
I've tried reading through what OpenSSH does, and honestly it's kind of a mess that I couldn't really hope to untangle in a reasonable amount of time.. :/